### ISSUE TITLE
Low FPS when using external display on hybrid NVIDIA laptop

### ISSUE THREAD

**Issue by marco-zan** [pop-os/cosmic-comp#928]
My hp laptop has this specs:
```
OS: Pop!_OS 22.04 LTS x86_64
Host: HP ENVY Laptop 15-ep0xxx
Kernel: 6.9.3-76060903-generic
Uptime: 13 mins
Packages: 3141 (dpkg), 109 (flatpak)
Shell: bash 5.1.16
Resolution: 3840x2160
CPU: Intel i7-10750H (12) @ 5.000GHz
GPU: Intel CometLake-H GT2 [UHD Graphics]
GPU: NVIDIA GeForce GTX 1650 Ti Mobile
Memory: 7544MiB / 15762MiB
```
Cosmic desktop has been installed following official methods, through apt.

Connected via HDMI to an external screen i notice that the framerate of the compositor, on this external screen, when doing (something, see below) would make the compositor lose some frames. While the internal screen just does perfectly his job. Note that this HDMI port is connected directly to the GPU

This is different from stuttering, because it still feels responsive and works, just there is the typical delay that you feel on a 30-40 ish FPS interface.

I fell like this happens everytime there are many "pixel that changes" on the screen. Like:
- Scrolling through the home of youtube (this is the most spot on example: not at all intensive, but with many images that changes)
- simply moving the cursor over a different windows (like firefox on the left and a terminal on the side)
- just scrolling through a terminal (with no programs running on it)

This is extremely better than anything i have seen on my external monitor using wayland, so please ask me if you need anything else.
This is the same experience i had on pop-os 24.04 alpha 1, when i tried it

**Comment by lucasslima** [comment]
This seems to affect all desktop environments when using a hybrid setup. Here's the thread on the nvidia forums: https://forums.developer.nvidia.com/t/nvidia-please-get-it-together-with-external-monitors-on-wayland/301684

Here's also a similar issue in plasma's kwin: https://bugs.kde.org/show_bug.cgi?id=452219

**Comment by Drakulix** [comment]
Yeah there is not much we can do here any more for the 1xxx-series.

Unfortunately this series has performance problems copying contents from the internal gpu to the external one, which we have to do depending on the gpu a specific application in question is using. We are already doing as many operations as possibly on the dedicated gpu to minimize latency.

So yes, the bigger the changes on screen the more needs to be copied, the worse the latency will be.

The only way around this is to do all the rendering on the nvidia gpu. You can force this by settings `COSMIC_RENDER_DEVICE` and pointing it to the render node that represents your nvidia gpu (e.g. `/dev/dri/renderD129`, but this depends on your system). This will likely fix performance on your external outputs, while the internal will most likely suffer now. (As will your battery, even if no external output is plugged in, as the dedicated gpu will be used by everything.)

You can also get around this by running a specific application in fullscreen on the nvidia gpu and it's external output. To do that use one of the special wayland sockets cosmic is providing (e.g. `WAYLAND_DISPLAY=wayland-1-renderD129`, note this doesn't work with Xwayland applications and we are working with upstream to make this work seamlessly without user-intervention).

As you can see we are doing what we can for the platform here. Everything else is a problem of the nvidia driver and/or gpu architecture (as 2xxx and later don't suffer from this to that extend).

**Comment by Drakulix** [comment]
Yes it is broader, but we have had cases where the 1xxx-series struggled to deliver anything comparable to 80/90 fps.

**Comment by marco-zan** [comment]
Thank you all guys. I didn't know at all about the thread on Nvidia forum, and about the options to change the default GPU on cosmic. It's awesome

**Comment by Drakulix** [comment]
> Sorry to bother you @Drakulix , I am wondering, how hard would it be to create a script, similar to system76-power, but just for switching between the integrated / dedicated graphics on wayland?

I am sorry, but this simply isn't how any of this works. The compositor has to partially use the GPUs the applications use and has to interact with the GPUs, that have the outputs connected.

Wayland has nothing to do with this, on X11 it was the same, just that the XServer managed this for you based on static configuration (which means this wasn't changeable after it was launched, which is why you had to log out on Pop!_OS with system76-power). This is simply not necessary on wayland. cosmic-comp will free the GPU, if nothing uses it.

> The rationale goes down to the fact that the dedicated gpu, if active but not in use, still draws more than 4 watts alone

Newer/modern cards shouldn't draw any power in this state and go into what is called "deep sleep". Some older generations will indeed not do this and still consume *some* power.

The problem is, that there exists no shared interface to turn these cards off. You usually have to do some completely undocumented ACPI calls to do that, depending on your machine.

For system76 laptops we made `system76-power` to change the XServer config, but *also* turn off the GPU correctly.

*If* you have an older System76 machine and you want to turn that GPU off, you can still use `system76-power`. The XServer files don't matter for COSMIC and the GPU will be turned off as usual, when switched to integrated.

For all other machines, you'll have to figure out manually how to turn it off and I can't help you with that, as this is very specific to a particular model.

> My question is, would be blacklisting the drivers enough to disable the dedicated card?

Denylisting the kernel modules won't do what you want. No loaded driver doesn't mean the card is turned off. Loaded drivers will just prevent you from deactivating the card, if you do know the magic command to turn it off.
